File: output/cpp/tests/51_hsm_persist.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


// Test 51: HSM Persistence (C++)
//
// Tests that HSM parent compartment chain is properly saved and restored:
// - Child state vars preserved
// - Parent state vars preserved
// - Parent compartment chain intact after restore
// - Forwarding to parent still works after restore

#include <iostream>
#include <string>
#include <cassert>
#include <nlohmann/json.hpp>

class HSMPersistFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    HSMPersistFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class HSMPersistFrameContext {
public:
    HSMPersistFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    HSMPersistFrameContext(HSMPersistFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class HSMPersistCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<HSMPersistFrameEvent> forward_event;
    std::unique_ptr<HSMPersistCompartment> parent_compartment;

    explicit HSMPersistCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<HSMPersistCompartment> clone() const {
        auto c = std::make_unique<HSMPersistCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class HSMPersist {
private:
    std::vector<std::unique_ptr<HSMPersistCompartment>> _state_stack;
    std::unique_ptr<HSMPersistCompartment> __compartment;
    std::unique_ptr<HSMPersistCompartment> __next_compartment;
    std::vector<HSMPersistFrameContext> _context_stack;

    void __kernel(HSMPersistFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            HSMPersistFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                HSMPersistFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    HSMPersistFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(HSMPersistFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Child") {
            _state_Child(__e);
        } else if (state_name == "Parent") {
            _state_Parent(__e);
        }
    }

    void __transition(std::unique_ptr<HSMPersistCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Parent(HSMPersistFrameEvent& __e) {
        auto* __sv_comp = __compartment.get();
        while (__sv_comp && __sv_comp->state != "Parent") { __sv_comp = __sv_comp->parent_compartment.get(); }
        if (__e._message == "$>") {
            if (__compartment->state_vars.count("parent_count") == 0) { __compartment->state_vars["parent_count"] = std::any(100); }
        } else if (__e._message == "get_parent_count") {
            _context_stack.back()._return = std::any(std::any_cast<int>(__sv_comp->state_vars["parent_count"]));
            return;
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("Parent"));
            return;
        } else if (__e._message == "increment_parent") {
            __sv_comp->state_vars["parent_count"] = std::any(std::any_cast<int>(__sv_comp->state_vars["parent_count"]) + 1);
        }
    }

    void _state_Child(HSMPersistFrameEvent& __e) {
        auto* __sv_comp = __compartment.get();
        while (__sv_comp && __sv_comp->state != "Child") { __sv_comp = __sv_comp->parent_compartment.get(); }
        if (__e._message == "$>") {
            if (__compartment->state_vars.count("child_count") == 0) { __compartment->state_vars["child_count"] = std::any(0); }
        } else if (__e._message == "get_child_count") {
            _context_stack.back()._return = std::any(std::any_cast<int>(__sv_comp->state_vars["child_count"]));
            return;
        } else if (__e._message == "get_parent_count") {
            // Forward to parent
            _state_Parent(__e);
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("Child"));
            return;
        } else if (__e._message == "increment_child") {
            __sv_comp->state_vars["child_count"] = std::any(std::any_cast<int>(__sv_comp->state_vars["child_count"]) + 1);
        } else if (__e._message == "increment_parent") {
            // Forward to parent
            _state_Parent(__e);
        }
    }

public:
    HSMPersist() {
        // HSM: Create parent compartment chain
        auto __parent_comp_0 = std::make_unique<HSMPersistCompartment>("Parent");
        __parent_comp_0->state_vars["parent_count"] = std::any(100);
        __compartment = std::make_unique<HSMPersistCompartment>("Child");
        __compartment->parent_compartment = std::move(__parent_comp_0);
        HSMPersistFrameEvent __frame_event("$>");
        HSMPersistFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void increment_child() {
        HSMPersistFrameEvent __e("increment_child");
        HSMPersistFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void increment_parent() {
        HSMPersistFrameEvent __e("increment_parent");
        HSMPersistFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    int get_child_count() {
        HSMPersistFrameEvent __e("get_child_count");
        HSMPersistFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    int get_parent_count() {
        HSMPersistFrameEvent __e("get_parent_count");
        HSMPersistFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::string get_state() {
        HSMPersistFrameEvent __e("get_state");
        HSMPersistFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::string save_state() {
        std::function<nlohmann::json(const HSMPersistCompartment*)> __ser = [&](const HSMPersistCompartment* c) -> nlohmann::json {
            if (!c) return nullptr;
            nlohmann::json __cj;
            __cj["state"] = c->state;
            nlohmann::json __sv;
            for (auto& [k, v] : c->state_vars) {
                if (k == "child_count") { try { __sv[k] = std::any_cast<int>(v); } catch(...) {} }
                if (k == "parent_count") { try { __sv[k] = std::any_cast<int>(v); } catch(...) {} }
            }
            __cj["state_vars"] = __sv;
            __cj["parent"] = __ser(c->parent_compartment.get());
            return __cj;
        };
        nlohmann::json __j;
        __j["_compartment"] = __ser(__compartment.get());
        nlohmann::json __stack = nlohmann::json::array();
        for (auto& c : _state_stack) { __stack.push_back(__ser(c.get())); }
        __j["_state_stack"] = __stack;
        return __j.dump();
    }

    static HSMPersist restore_state(const std::string& json) {
        std::function<std::unique_ptr<HSMPersistCompartment>(const nlohmann::json&)> __deser = [&](const nlohmann::json& d) -> std::unique_ptr<HSMPersistCompartment> {
            if (d.is_null()) return nullptr;
            auto c = std::make_unique<HSMPersistCompartment>(std::string(d["state"]));
            if (d.contains("state_vars")) {
                auto& sv = d["state_vars"];
                if (sv.contains("child_count")) { c->state_vars["child_count"] = std::any(sv["child_count"].get<int>()); }
                if (sv.contains("parent_count")) { c->state_vars["parent_count"] = std::any(sv["parent_count"].get<int>()); }
            }
            if (d.contains("parent") && !d["parent"].is_null()) {
                c->parent_compartment = __deser(d["parent"]);
            }
            return c;
        };
        auto __j = nlohmann::json::parse(json);
        HSMPersist __instance;
        __instance.__compartment = __deser(__j["_compartment"]);
        if (__j.contains("_state_stack")) {
            for (auto& __sc : __j["_state_stack"]) {
                __instance._state_stack.push_back(__deser(__sc));
            }
        }
        return __instance;
    }
};

int main() {
    printf("=== Test 51: HSM Persistence (C++) ===\n");

    // Create system and modify state vars at both levels
    HSMPersist s1;

    // Verify initial state
    if (s1.get_state() != "Child") {
        printf("FAIL: Expected Child, got %s\n", s1.get_state().c_str());
        assert(false);
    }
    if (s1.get_child_count() != 0) {
        printf("FAIL: Expected 0, got %d\n", s1.get_child_count());
        assert(false);
    }
    if (s1.get_parent_count() != 100) {
        printf("FAIL: Expected 100, got %d\n", s1.get_parent_count());
        assert(false);
    }
    printf("1. Initial state verified: Child with child_count=0, parent_count=100\n");

    // Modify state vars at both levels
    s1.increment_child();
    s1.increment_child();
    s1.increment_child();  // child_count = 3
    s1.increment_parent();
    s1.increment_parent();  // parent_count = 102

    if (s1.get_child_count() != 3) {
        printf("FAIL: Expected 3, got %d\n", s1.get_child_count());
        assert(false);
    }
    if (s1.get_parent_count() != 102) {
        printf("FAIL: Expected 102, got %d\n", s1.get_parent_count());
        assert(false);
    }
    printf("2. After increments: child_count=%d, parent_count=%d\n", s1.get_child_count(), s1.get_parent_count());

    // Save state
    std::string json = s1.save_state();
    printf("3. Saved state: %zu chars\n", json.length());

    // Restore to new instance
    HSMPersist s2 = HSMPersist::restore_state(json);

    // Verify restored state
    if (s2.get_state() != "Child") {
        printf("FAIL: Expected Child after restore, got %s\n", s2.get_state().c_str());
        assert(false);
    }
    printf("4. Restored state: %s\n", s2.get_state().c_str());

    // Verify child state var preserved
    if (s2.get_child_count() != 3) {
        printf("FAIL: Expected child_count=3, got %d\n", s2.get_child_count());
        assert(false);
    }
    printf("5. Child state var preserved: child_count=%d\n", s2.get_child_count());

    // Verify parent state var preserved (requires parent compartment chain)
    if (s2.get_parent_count() != 102) {
        printf("FAIL: Expected parent_count=102, got %d\n", s2.get_parent_count());
        assert(false);
    }
    printf("6. Parent state var preserved: parent_count=%d\n", s2.get_parent_count());

    // Verify state machine still works after restore
    s2.increment_child();
    s2.increment_parent();
    if (s2.get_child_count() != 4) {
        printf("FAIL: Expected 4, got %d\n", s2.get_child_count());
        assert(false);
    }
    if (s2.get_parent_count() != 103) {
        printf("FAIL: Expected 103, got %d\n", s2.get_parent_count());
        assert(false);
    }
    printf("7. After post-restore increments: child_count=%d, parent_count=%d\n", s2.get_child_count(), s2.get_parent_count());

    printf("PASS: HSM persistence works correctly\n");
    return 0;
}

File: output/javascript/tests/51_hsm_persist.js

// Test 51: HSM Persistence
//
// Tests that HSM parent compartment chain is properly saved and restored:
// - Child state vars preserved
// - Parent state vars preserved
// - Parent compartment chain intact after restore
// - Forwarding to parent still works after restore

export class HSMPersistFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class HSMPersistFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class HSMPersistCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new HSMPersistCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class HSMPersist {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        // HSM: Create parent compartment chain
        const __parent_comp_0 = new HSMPersistCompartment("Parent", null);
        __parent_comp_0.state_vars["parent_count"] = 100;
        this.__compartment = new HSMPersistCompartment("Child", __parent_comp_0);
        this.__next_compartment = null;
        const __frame_event = new HSMPersistFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new HSMPersistFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new HSMPersistFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new HSMPersistFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    increment_child() {
        const __e = new HSMPersistFrameEvent("increment_child", null);
        const __ctx = new HSMPersistFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    increment_parent() {
        const __e = new HSMPersistFrameEvent("increment_parent", null);
        const __ctx = new HSMPersistFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    get_child_count() {
        const __e = new HSMPersistFrameEvent("get_child_count", null);
        const __ctx = new HSMPersistFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_parent_count() {
        const __e = new HSMPersistFrameEvent("get_parent_count", null);
        const __ctx = new HSMPersistFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_state() {
        const __e = new HSMPersistFrameEvent("get_state", null);
        const __ctx = new HSMPersistFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_Parent(__e) {
        // HSM: Navigate to this state's compartment for state var access
        let __sv_comp = this.__compartment;
        while (__sv_comp !== null && __sv_comp.state !== "Parent") {
            __sv_comp = __sv_comp.parent_compartment;
        }
        if (__e._message === "$>") {
            if (!("parent_count" in __sv_comp.state_vars)) {
                __sv_comp.state_vars["parent_count"] = 100;
            }
        } else if (__e._message === "get_parent_count") {
            this._context_stack[this._context_stack.length - 1]._return = __sv_comp.state_vars["parent_count"];
            return;
        } else if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Parent";
            return;
        } else if (__e._message === "increment_parent") {
            __sv_comp.state_vars["parent_count"] = __sv_comp.state_vars["parent_count"] + 1;
        }
    }

    _state_Child(__e) {
        // HSM: Navigate to this state's compartment for state var access
        let __sv_comp = this.__compartment;
        while (__sv_comp !== null && __sv_comp.state !== "Child") {
            __sv_comp = __sv_comp.parent_compartment;
        }
        if (__e._message === "$>") {
            if (!("child_count" in __sv_comp.state_vars)) {
                __sv_comp.state_vars["child_count"] = 0;
            }
        } else if (__e._message === "get_child_count") {
            this._context_stack[this._context_stack.length - 1]._return = __sv_comp.state_vars["child_count"];
            return;
        } else if (__e._message === "get_parent_count") {
            // Forward to parent
            this._state_Parent(__e);
        } else if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Child";
            return;
        } else if (__e._message === "increment_child") {
            __sv_comp.state_vars["child_count"] = __sv_comp.state_vars["child_count"] + 1;
        } else if (__e._message === "increment_parent") {
            // Forward to parent
            this._state_Parent(__e);
        }
    }

    saveState() {
        const serializeComp = (c) => {
            if (!c) return null;
            return {
                state: c.state,
                state_args: {...c.state_args},
                state_vars: {...c.state_vars},
                enter_args: {...c.enter_args},
                exit_args: {...c.exit_args},
                forward_event: c.forward_event,
                parent_compartment: serializeComp(c.parent_compartment),
            };
        };
        return JSON.stringify({
            _compartment: serializeComp(this.__compartment),
            _state_stack: this._state_stack.map((c) => serializeComp(c)),
        });
    }

    static restoreState(json) {
        const deserializeComp = (data) => {
            if (!data) return null;
            const comp = new HSMPersistCompartment(data.state);
            comp.state_args = {...(data.state_args || {})};
            comp.state_vars = {...(data.state_vars || {})};
            comp.enter_args = {...(data.enter_args || {})};
            comp.exit_args = {...(data.exit_args || {})};
            comp.forward_event = data.forward_event;
            comp.parent_compartment = deserializeComp(data.parent_compartment);
            return comp;
        };
        const data = JSON.parse(json);
        const instance = Object.create(HSMPersist.prototype);
        instance.__compartment = deserializeComp(data._compartment);
        instance.__next_compartment = null;
        instance._state_stack = (data._state_stack || []).map((c) => deserializeComp(c));
        instance._context_stack = [];
        return instance;
    }
}

function main() {
    console.log("=== Test 51: HSM Persistence ===");

    // Create system and modify state vars at both levels
    const s1 = new HSMPersist();

    // Verify initial state
    console.assert(s1.get_state() === "Child", `Expected Child, got ${s1.get_state()}`);
    console.assert(s1.get_child_count() === 0, `Expected 0, got ${s1.get_child_count()}`);
    console.assert(s1.get_parent_count() === 100, `Expected 100, got ${s1.get_parent_count()}`);
    console.log("1. Initial state verified: Child with child_count=0, parent_count=100");

    // Modify state vars at both levels
    s1.increment_child();
    s1.increment_child();
    s1.increment_child();  // child_count = 3
    s1.increment_parent();
    s1.increment_parent();  // parent_count = 102

    console.assert(s1.get_child_count() === 3, `Expected 3, got ${s1.get_child_count()}`);
    console.assert(s1.get_parent_count() === 102, `Expected 102, got ${s1.get_parent_count()}`);
    console.log(`2. After increments: child_count=${s1.get_child_count()}, parent_count=${s1.get_parent_count()}`);

    // Save state
    const data = s1.saveState();
    console.assert(typeof data === "string", `Expected string, got ${typeof data}`);
    console.log(`3. Saved state: ${data.length} chars`);

    // Restore to new instance
    const s2 = HSMPersist.restoreState(data);

    // Verify restored state
    console.assert(s2.get_state() === "Child", `Expected Child after restore, got ${s2.get_state()}`);
    console.log(`4. Restored state: ${s2.get_state()}`);

    // Verify child state var preserved
    console.assert(s2.get_child_count() === 3, `Expected child_count=3, got ${s2.get_child_count()}`);
    console.log(`5. Child state var preserved: child_count=${s2.get_child_count()}`);

    // Verify parent state var preserved (requires parent compartment chain)
    console.assert(s2.get_parent_count() === 102, `Expected parent_count=102, got ${s2.get_parent_count()}`);
    console.log(`6. Parent state var preserved: parent_count=${s2.get_parent_count()}`);

    // Verify state machine still works after restore
    s2.increment_child();
    s2.increment_parent();
    console.assert(s2.get_child_count() === 4, `Expected 4, got ${s2.get_child_count()}`);
    console.assert(s2.get_parent_count() === 103, `Expected 103, got ${s2.get_parent_count()}`);
    console.log(`7. After post-restore increments: child_count=${s2.get_child_count()}, parent_count=${s2.get_parent_count()}`);

    console.log("PASS: HSM persistence works correctly");
}

main();
